File: 11-cosmic-expansion/app/Main.hs
{-# LANGUAGE RecordWildCards #-}
{-# OPTIONS_GHC -Wno-name-shadowing #-}

module Main (main) where

import           Data.List (elemIndices, tails, (\\))

type Step = Int
type Coords = (Int, Int)  -- row, col
type Galaxies = [Coords]

data Image = Image { galaxies :: Galaxies
                   , width    :: Int
                   , height   :: Int
                   } deriving (Show, Eq)

parseImage :: String -> Image
parseImage s = let enumerated = zip [0..] (lines s)
                in Image{ galaxies = [(row, col) | (row, line) <- enumerated, col <- parseLine line]
                        , width = length (head (lines s))
                        , height = length (lines s)
                        }

parseLine :: String -> [Int]
parseLine = elemIndices '#'

galaxyPairs :: Image -> [(Coords, Coords)]
galaxyPairs Image{..} = [(x, y) | (x:rest) <- tails galaxies, y <- rest]

expand :: Image -> Image
expand i@Image{..} = expandCols emptyCols $ expandRows emptyRows i
  where expandCols :: [Int] -> Image -> Image
        expandCols xs n = foldr expandCol n xs

        expandCol :: Int -> Image -> Image
        -- part 1
        -- expandCol n Image{..} = Image{galaxies=fmap (\(row, col) -> if col > n then (row, col + 1) else (row, col)) galaxies, width=width+1, height=height}
        -- part 2
        expandCol n Image{..} = Image{galaxies=fmap (\(row, col) -> if col > n then (row, col + million) else (row, col)) galaxies, width=width+million, height=height}

        expandRows :: [Int] -> Image -> Image
        expandRows xs n = foldr expandRow n xs

        expandRow :: Int -> Image -> Image
        -- part 1
        -- expandRow n Image{..} = Image{galaxies=fmap (\(row, col) -> if row > n then (row + 1, col) else (row, col)) galaxies, width=width, height=height+1}
        -- part 2
        expandRow n Image{..} = Image{galaxies=fmap (\(row, col) -> if row > n then (row + million, col) else (row, col)) galaxies, width=width, height=height+million}

        million = 1000000 - 1  -- ;p
        emptyRows = [0 .. width - 1] \\ map fst galaxies
        emptyCols = [0 .. height - 1] \\ map snd galaxies


distBtw :: Coords -> Coords -> Step
distBtw (x1, y1) (x2, y2) = abs (x1 - x2) + abs (y1 - y2)

main :: IO ()
main = do input <- readFile "input.txt"
          let image = parseImage input
          let expanded = expand image
          let gPairs = galaxyPairs expanded
          let dists = map (uncurry distBtw) gPairs
          let total = sum dists
          print total

File: 11-cosmic-expansion/main.py
# pyright: strict

from typing import Iterable, Set, Tuple
from dataclasses import dataclass, replace
from functools import reduce


Coords = Tuple[int, int]


@dataclass
class Image:
    galaxies: Set[Coords]
    height: int
    width: int


def parse_image(txt: str) -> Image:
    lines = txt.splitlines()
    width = len(lines[0])
    height = len(lines)
    return Image(galaxies={ (row, idx) for (row, line) in enumerate(lines)
                                       for idx in galaxy_indices(line) },
                 width=width,
                 height=height)


def galaxy_indices(line: str) -> Set[int]:
    return { i for (i, c) in enumerate(line) if c == '#' }


def dist_btw(g1: Coords, g2: Coords) -> int:
    (x1, y1), (x2, y2) = g1, g2
    return abs(x1 - x2) + abs(y1 - y2)


def is_empty_row(image: Image, row: int) -> bool:
    return row not in { r for (r, _) in image.galaxies }


def is_empty_col(image: Image, col: int) -> bool:
    return col not in { c for (_, c) in image.galaxies }


# expand empty rows in the image, substituting them by <factor> empty rows
def expand_horizontal(image: Image, factor: int) -> Image:
    empty_rows = { row for row in range(0, image.height) if is_empty_row(image, row) }
    descending = sorted(empty_rows, reverse=True)

    def expand_one_row(image: Image, row: int) -> Image:
        affected = { (r, c) for (r, c) in image.galaxies if r > row }
        stationary = image.galaxies - affected
        moved = { (r + factor - 1, c) for (r, c) in affected }
        return replace(image, galaxies=stationary | moved)

    return reduce(expand_one_row, descending, image)


# expand empty columns in the image, substituting them by <factor> empty columns
def expand_vertical(image: Image, factor: int) -> Image:
    empty_cols = { col for col in range(0, image.width) if is_empty_col(image, col) }
    descending = sorted(empty_cols, reverse=True)

    def expand_one_col(image: Image, col: int) -> Image:
        affected = { (r, c) for (r, c) in image.galaxies if c > col }
        stationary = image.galaxies - affected
        moved = { (r, c + factor - 1) for (r, c) in affected }
        return replace(image, galaxies=stationary | moved)

    return reduce(expand_one_col, descending, image)


def expand(image: Image, factor: int) -> Image:
    return expand_vertical(expand_horizontal(image, factor), factor)


Pair = Tuple[Coords, Coords]
def galaxy_pairs(image: Image) -> Iterable[Pair]:
    return { (g1, g2) for g1 in image.galaxies
                      for g2 in image.galaxies
                      if g1 < g2 }


def main():
    handle = open('input.txt', 'r')
    text = handle.read()
    handle.close()
    image = parse_image(text)
    expanded = expand(image, factor=2)
    pairs = galaxy_pairs(expanded)
    dists = [dist_btw(g1, g2) for g1, g2 in pairs]
    total = sum(dists)
    print(total)


if __name__ == '__main__':
    main()
